File: 3. Behavioral Design Patterns/Visitor/Python/VisitorTemplate.py
# Visitor Pattern – Python Template

from abc import ABC, abstractmethod

class Visitor(ABC):
    @abstractmethod
    def visit_concrete_element_a(self, element):
        pass
    
    @abstractmethod
    def visit_concrete_element_b(self, element):
        pass

class Element(ABC):
    @abstractmethod
    def accept(self, visitor):
        pass

class ConcreteElementA(Element):
    def accept(self, visitor):
        visitor.visit_concrete_element_a(self)
    
    def operation_a(self):
        return "ConcreteElementA operation"

class ConcreteElementB(Element):
    def accept(self, visitor):
        visitor.visit_concrete_element_b(self)
    
    def operation_b(self):
        return "ConcreteElementB operation"

class ConcreteVisitor1(Visitor):
    def visit_concrete_element_a(self, element):
        print(f"ConcreteVisitor1: {element.operation_a()}")
    
    def visit_concrete_element_b(self, element):
        print(f"ConcreteVisitor1: {element.operation_b()}")

class ConcreteVisitor2(Visitor):
    def visit_concrete_element_a(self, element):
        print(f"ConcreteVisitor2: {element.operation_a()}")
    
    def visit_concrete_element_b(self, element):
        print(f"ConcreteVisitor2: {element.operation_b()}")

def main():
    elements = [ConcreteElementA(), ConcreteElementB()]
    visitor = ConcreteVisitor1()
    
    for element in elements:
        element.accept(visitor)

if __name__ == "__main__":
    main()

File: 3. Behavioral Design Patterns/Visitor/Kotlin/VisitorTemplate.kt
// Visitor Pattern – Kotlin Template

interface Visitor {
    fun visit(element: ConcreteElementA)
    fun visit(element: ConcreteElementB)
}

interface Element {
    fun accept(visitor: Visitor)
}

class ConcreteElementA : Element {
    override fun accept(visitor: Visitor) = visitor.visit(this)
    fun operationA() { /* specific to A */ }
}

class ConcreteElementB : Element {
    override fun accept(visitor: Visitor) = visitor.visit(this)
    fun operationB() { /* specific to B */ }
}

class ConcreteVisitor1 : Visitor {
    override fun visit(element: ConcreteElementA) { element.operationA(); /* visitor 1 logic */ }
    override fun visit(element: ConcreteElementB) { element.operationB(); /* visitor 1 logic */ }
}

class ConcreteVisitor2 : Visitor {
    override fun visit(element: ConcreteElementA) { element.operationA(); /* visitor 2 logic */ }
    override fun visit(element: ConcreteElementB) { element.operationB(); /* visitor 2 logic */ }
}

fun main() {
    val elements: Array<Element> = arrayOf(ConcreteElementA(), ConcreteElementB())
    val visitor = ConcreteVisitor1()
    for (element in elements) element.accept(visitor)
}
